### PR TITLE
Adds `export-state` subcommand

### PR DESCRIPTION
**Pull request by bkchr**
This adds a new sub-command `export-state`. This sub-command can be used to export the state of a running chain as a chain spec. This is useful for testing runtime upgrades before distributing them on a live chain.

It basically collects all the key value pairs at a given block and overwrites the genesis storage values with the collected ones. The build chain-spec will not work right away, because `Babe` will complain about unexpected epoch changes and the authority set will contain authorities for which the user doesn't have the key. This will be solved by an extra tool that replaces the corresponding keys and makes the chain spec usable as new a new genesis.

### CODE REVIEW

File: client/service/src/chain_ops.rs [hunk 1]
```diff
@@ -298,4 +299,45 @@ impl<
...
+			|| BlockId::Hash(self.client.usage_info().chain.best_hash)
+		);
+
+		let key = StorageKey(Vec::new());
+		let mut top_storage = self.client.storage_pairs(&block, &key)?;
+		let mut children_default = HashMap::new();
+
+		// Remove all default child storage roots from the top storage and collect the child storage
+		// pairs.
+		while let Some(pos) = top_storage
```
Comment (bkchr) [inline client/service/src/chain_ops.rs:317]: @cheme it would be nice if you could take a look here, if I have done this child value collection correctly.

**Comment by arkpar** [comment]
Looks good, except it loads all of the state in memory. This is fine for now, but might cause problems with bigger chains in the future.
Ideally it should be writing the file while iterating over the state.

**Comment by bkchr** [comment]
> Looks good, except it loads all of the state into memory. This is fine for now, but might cause problems with bigger chains in the future.
> Ideally it should be writing the file while iterating over the state.

Yeah, the whole operation isn't that nice, but as this is only some sort of debug command. I thought that we can live with the high memory load. ;)

File: client/service/src/chain_ops.rs [hunk 2]
```diff
@@ -298,4 +299,45 @@ impl<
...
+			let mut pairs = StorageMap::new();
+			keys.into_iter().try_for_each(|k| {
+				if let Some(value) = self.client.child_storage(&block, &child_info, &k)? {
+					pairs.insert(k.0, value.0);
+				}
+
+				Ok::<_, Error>(())
+			})?;
+
+			children_default.insert(key.0, StorageChild { child_info, data: pairs });
```
Comment (cheme) [inline client/service/src/chain_ops.rs:337]: Here we should insert with key (without prefix), children_default works on unprefixed key.
If only use at https://github.com/paritytech/substrate/blob/eb259b2f0156ac14c956b374f2e9033b21984448/client/db/src/lib.rs#L637 it is ignored, but could be an issue later.

File: primitives/storage/src/lib.rs [hunk 3]
```diff
@@ -148,6 +148,9 @@ pub mod well_known_keys {
 	/// Prefix of child storage keys.
 	pub const CHILD_STORAGE_KEY_PREFIX: &'static [u8] = b":child_storage:";
 
+	/// Prefix of the default child storage keys in the top trie.
+	pub const DEFAULT_CHILD_STORAGE_KEY_PREFIX: &'static [u8] = b":child_storage:default:";
```
Comment (cheme) [inline primitives/storage/src/lib.rs:152]: Ideally this should not be made public, we would use
```
let (type, key) = ChildType::from_prefixed_key(prefixed_key);
assert!(type == ChildType::Default);
```

File: client/service/src/chain_ops.rs (hunk 2 above)
Comment (cheme) [reply client/service/src/chain_ops.rs:337]: Ok I m stoopid, key was redefined with the unprefixed one so it is the unprefixed one, so please forget previous comment.
Algo looks correct to :+1:

File: client/service/src/chain_ops.rs [hunk 4]
```diff
@@ -298,4 +299,45 @@ impl<
...
+			.iter()
+			.position(|(k, _)| k.0.starts_with(well_known_keys::DEFAULT_CHILD_STORAGE_KEY_PREFIX)) {
+			let (key, _) = top_storage.swap_remove(pos);
+
+			let key = StorageKey(
+				key.0[well_known_keys::DEFAULT_CHILD_STORAGE_KEY_PREFIX.len()..].to_vec(),
+			);
+			let child_info = ChildInfo::new_default(&key.0);
+
+			let keys = self.client.child_storage_keys(&block, &child_info, &key)?;
```
Comment (cheme) [inline client/service/src/chain_ops.rs:327]: Here third parameter should be empty prefix not key.

File: client/service/src/chain_ops.rs [hunk 5]
```diff
@@ -298,4 +299,45 @@ impl<
...
+
+	fn export_raw_state(
+		&self,
+		block: Option<BlockId<Self::Block>>,
+	) -> Result<Storage, Error> {
+		let block = block.unwrap_or_else(
+			|| BlockId::Hash(self.client.usage_info().chain.best_hash)
+		);
+
+		let key = StorageKey(Vec::new());
```
Comment (bkchr) [inline client/service/src/chain_ops.rs:311]:
```suggestion
		let empty_key = StorageKey(Vec::new());
```

File: client/service/src/chain_ops.rs [hunk 6]
```diff
@@ -298,4 +299,45 @@ impl<
...
+		let mut top_storage = self.client.storage_pairs(&block, &key)?;
```
Comment (bkchr) [inline client/service/src/chain_ops.rs:312]:
```suggestion
		let mut top_storage = self.client.storage_pairs(&block, &empty_key)?;
```

File: client/service/src/chain_ops.rs (hunk 4 above)
Comment (bkchr) [reply client/service/src/chain_ops.rs:327]: Good find ;)

Comment (bkchr) [inline client/service/src/chain_ops.rs:327]:
```suggestion
			let keys = self.client.child_storage_keys(&block, &child_info, &empty_key)?;
```

File: client/cli/src/commands/check_block_cmd.rs [hunk 7]
```diff
@@ -14,23 +14,20 @@
...
+use std::{fmt::Debug, str::FromStr};
 use structopt::StructOpt;
 
 /// The `check-block` command used to validate blocks.
 #[derive(Debug, StructOpt, Clone)]
 pub struct CheckBlockCmd {
 	/// Block hash or number
 	#[structopt(value_name = "HASH or NUMBER")]
-	pub input: String,
+	pub input: BlockNumberOrHash,
```
Comment (cecton) [inline client/cli/src/commands/check_block_cmd.rs:30]: Sweet!

File: client/cli/src/params/mod.rs [hunk 8]
```diff
@@ -64,8 +64,89 @@ impl BlockNumber {
...
+
+	fn from_str(block_number: &str) -> Result<Self, Self::Err> {
+		if block_number.starts_with("0x") {
+			if let Some(pos) = &block_number[2..].chars().position(|c| !c.is_ascii_hexdigit()) {
+				Err(format!(
+					"Expected block hash, found illegal hex character at position: {}",
+					2 + pos,
+				))
+			} else {
+				Ok(Self(block_number.into()))
```
Comment (cecton) [inline client/cli/src/params/mod.rs:86]: It would be probably handy if we had a `try_into` for this

**Comment by xlc** [comment]
Nice. Where can I find this "extra tool that replaces the corresponding keys and makes the chain spec usable as new a new genesis" ?

**Comment by bkchr** [comment]
This does not exist yet, I will work on this in the next time.

**Comment by xcaptain** [comment]
@bkchr I use `export-state` to create a new genesis file and use `--chain state_file.json` to start a new chain, I also got the `ClientImport(expected epoch change)` error, is that because my old keys changed?

**Pull request merged**